webapp/test_results_handler.go

```diff
@@ -42,122 +53,106 @@ func testResultsHandler(w http.ResponseWriter, r *http.Request) {
+		TestRunFilter testRunsUIFilter
 	}{
-		SHA:    runSHA,
-		Filter: r.URL.Query().Get("filter"),
+		TestRunFilter: filter,
 	}
 
-	labels := shared.ToStringSlice(shared.ParseLabelsParam(r))
-	if labels != nil {
-		var marshaled []byte
-		if marshaled, err = json.Marshal(labels); err != nil {
-			http.Error(w, err.Error(), http.StatusInternalServerError)
-			return
-		}
-		data.Labels = string(marshaled)
-	}
-
-	_, diff := query["diff"]
-	data.Diff = diff || query.Get("before") != "" || query.Get("after") != ""
-
-	// Run source URLs
-	if testRunSources != nil && len(testRunSources) > 0 {
-		var marshaled []byte
-		if marshaled, err = json.Marshal(testRunSources); err != nil {
-			http.Error(w, err.Error(), http.StatusInternalServerError)
-			return
+	// Runs by base64-encoded param or spec param.
+	if filter.BeforeTestRun != nil && filter.AfterTestRun != nil {
+		runs := []shared.TestRun{
+			*(filter.BeforeTestRun),
+			*(filter.AfterTestRun),
 		}
-		data.TestRunSources = string(marshaled)
-	}
+		filter.Diff = true
 
-	// Runs by base64-encoded param or spec param.
-	if testRuns != nil && len(testRuns) > 0 {
 		var marshaled []byte
-		if marshaled, err = json.Marshal(testRuns); err != nil {
+		if marshaled, err = json.Marshal(runs); err != nil {
 			http.Error(w, err.Error(), http.StatusInternalServerError)
 			return
 		}
 		data.TestRuns = string(marshaled)
 	}
 
-	if err := templates.ExecuteTemplate(w, "index.html", data); err != nil {
+	if err := templates.ExecuteTemplate(w, "results.html", data); err != nil {
 		http.Error(w, err.Error(), http.StatusInternalServerError)
 		return
 	}
 }
 
-// getTestRunsAndSources gets source urls (api endpoints), and any placeholder TestRun entities from the parameters
-// for the current request.
-// When diffing, 'before' and 'after' parameters can be test-run specs (i.e. [product]@[sha]), or base64 encoded
-// TestRun JSON blobs for the results summaries.
-func getTestRunsAndSources(r *http.Request) (testRunSources []string, testRuns []shared.TestRun, err error) {
+// parseTestRunUIFilter parses the standard TestRunFilter, as well as the extra
+// diff params (diff, before, after).
+func parseTestRunUIFilter(r *http.Request) (filter testRunsUIFilter, err error) {
+	testRunFilter, err := shared.ParseTestRunFilterParams(r)
+	if err != nil {
+		return filter, err
+	}
+
 	before := r.URL.Query().Get("before")
 	after := r.URL.Query().Get("after")
 	if before != "" || after != "" {
 		if before == "" {
-			return nil, nil, errors.New("after param provided, but before param missing")
+			return filter, errors.New("after param provided, but before param missing")
 		} else if after == "" {
-			return nil, nil, errors.New("before param provided, but after param missing")
+			return filter, errors.New("before param provided, but after param missing")
 		}
 
 		const singleRunURL = `/api/run?sha=%s&product=%s`
 
 		if beforeDecoded, err := base64.URLEncoding.DecodeString(before); err == nil {
```

Comment:
Offline chat with Luke: so it's only the client-side base64 transformation that's being removed; the backend still supports base64 params so that old URLs still work.